quickwit/quickwit-actors/src/kill_switch.rs

```diff
@@ -18,36 +18,82 @@
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
 use std::sync::atomic::{AtomicBool, Ordering};
-use std::sync::Arc;
+use std::sync::{Arc, Mutex, Weak};
 
 use tracing::debug;
 
-#[derive(Clone)]
+#[derive(Clone, Default)]
 pub struct KillSwitch {
-    alive: Arc<AtomicBool>,
+    inner: Arc<Inner>,
 }
 
-impl Default for KillSwitch {
+struct Inner {
+    alive: AtomicBool,
+    children: Mutex<Vec<Weak<Inner>>>,
+}
+
+impl Default for Inner {
     fn default() -> Self {
-        KillSwitch {
-            alive: Arc::new(AtomicBool::new(true)),
+        Self {
+            alive: AtomicBool::new(true),
+            children: Mutex::default(),
         }
     }
 }
 
-impl KillSwitch {
-    pub fn kill(&self) {
-        debug!("kill-switch-activated");
-        self.alive.store(false, Ordering::Relaxed);
+fn gc(children: &mut Vec<Weak<Inner>>) {
```

Comment:
nitpick: use `garbage_collect` as we tend to use longer and explicit names in the project.